pkg/controller.v1/jax/jaxjob_controller.go

```diff
@@ -0,0 +1,485 @@
+	}
+
+	// eventHandler for owned object
+	eventHandler := handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &kubeflowv1.JAXJob{}, handler.OnlyControllerOwner())
+	predicates := predicate.Funcs{
+		CreateFunc: util.OnDependentCreateFunc(r.Expectations),
+		UpdateFunc: util.OnDependentUpdateFunc(&r.JobController),
+		DeleteFunc: util.OnDependentDeleteFunc(r.Expectations),
+	}
+	// Create generic predicates
+	genericPredicates := predicate.Funcs{
+		CreateFunc: util.OnDependentCreateFuncGeneric(r.Expectations),
+		UpdateFunc: util.OnDependentUpdateFuncGeneric(&r.JobController),
+		DeleteFunc: util.OnDependentDeleteFuncGeneric(r.Expectations),
+	}
+	// inject watching for job related pod
+	if err = c.Watch(source.Kind(mgr.GetCache(), &corev1.Pod{}), eventHandler, predicates); err != nil {
+		return err
+	}
+	// inject watching for job related service
+	if err = c.Watch(source.Kind(mgr.GetCache(), &corev1.Service{}), eventHandler, predicates); err != nil {
+		return err
+	}
+	// skip watching volcano PodGroup if volcano PodGroup is not installed
+	if _, err = mgr.GetRESTMapper().RESTMapping(schema.GroupKind{Group: v1beta1.GroupName, Kind: "PodGroup"},
+		v1beta1.SchemeGroupVersion.Version); err == nil {
+		// inject watching for job related volcano PodGroup
+		if err = c.Watch(source.Kind(mgr.GetCache(), &v1beta1.PodGroup{}), eventHandler, genericPredicates); err != nil {
+			return err
+		}
+	}
+	// skip watching scheduler-plugins PodGroup if scheduler-plugins PodGroup is not installed
+	if _, err = mgr.GetRESTMapper().RESTMapping(schema.GroupKind{Group: schedulerpluginsv1alpha1.SchemeGroupVersion.Group, Kind: "PodGroup"},
+		schedulerpluginsv1alpha1.SchemeGroupVersion.Version); err == nil {
+		// inject watching for job related scheduler-plugins PodGroup
+		if err = c.Watch(source.Kind(mgr.GetCache(), &schedulerpluginsv1alpha1.PodGroup{}), eventHandler, genericPredicates); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *JAXJobReconciler) ControllerName() string {
+	return controllerName
+}
+
+func (r *JAXJobReconciler) GetAPIGroupVersionKind() schema.GroupVersionKind {
+	return kubeflowv1.GroupVersion.WithKind(kubeflowv1.JAXJobKind)
+}
+
+func (r *JAXJobReconciler) GetAPIGroupVersion() schema.GroupVersion {
+	return kubeflowv1.GroupVersion
+}
+
+func (r *JAXJobReconciler) GetGroupNameLabelValue() string {
+	return kubeflowv1.GroupVersion.Group
+}
+
+func (r *JAXJobReconciler) GetFrameworkName() string {
+	return kubeflowv1.JAXJobFrameworkName
+}
+
+func (r *JAXJobReconciler) GetJobFromInformerCache(namespace, name string) (metav1.Object, error) {
+	job := &kubeflowv1.JAXJob{}
+	err := r.Get(context.Background(), types.NamespacedName{Namespace: namespace, Name: name}, job)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logrus.Error(err, "jax job not found", "namespace", namespace, "name", name)
+		} else {
+			logrus.Error(err, "failed to get job from api-server", "namespace", namespace, "name", name)
+		}
```

Comment:
We can report error directly instead of two different errors?

Reply:
Got this from existing controllers implementation